### PR TITLE
fix(ios): deletion of selected text

### LINKED ISSUE #11095: bug(ios): No space between words when replacing selected text with a letter
### Describe the bug

I noticed that after pressing a letter to replace selected text, there is no space between the words.

### Reproduce the bug

1. Install Keyman 17.0.295-beta
2. Open Keyman app.
3. Type two words. eg., Something Beautiful
4. Select the second word 'beautiful'.
5. Press letter 'B'. 

Here, I observed that there is no space between the word 'Something' and the replaced text 'B'. Seems to be an issue.

Note: I am not able to reproduce this issue in Android OS. 

I have attached the Video file for reference. 

..no space

https://github.com/keymanapp/keyman/assets/19683143/4f44456c-0736-4c82-bb25-f7be7cab1897

### Expected behavior

There should be a space between two words after pressing any key to the selected (second) word. 

### Related issues

_No response_

### Keyman apps

- [ ] Keyman for Android
- [X] Keyman for iPhone and iPad
- [ ] Keyman for Linux
- [ ] Keyman for macOS
- [ ] Keyman for Windows
- [ ] Keyman Developer
- [ ] KeymanWeb
- [ ] Other - give details at bottom of form

### Keyman version

17.0.295-beta

### Operating system

iOS 17.4, iOS 15.5

### Device

iPhone 13 Mobile

### Target application

_No response_

### Browser

_No response_

### Keyboard name

_No response_

### Keyboard version

_No response_

### Language name

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jahorton**
Fixes #11095.

## User Testing

TEST_SELECTION_REPLACEMENT: Attempt to repro #11095.

1. Install the Keyman for iOS version associated with this PR for user testing.
2. Type `Something beautiful`  - it's fine to use predictive-text suggestions to help
3. Select the second word: 'beautiful'.
4. Press letter 'B'. 

Expected result: `Something B`, with a space between `Something` and `B`.

TEST_SELECTION_BACKSPACE:  Using the iOS Keyman app, verify that pressing backspace to delete a selection acts as intended - the selected text should be removed, with no further edits.  Any preceding spaces should remain.

TEST_SELECTION_REPLACEMENT_MID:  Replace selected text that starts in the middle of a word.
1. Install the Keyman for iOS version associated with this PR for user testing.
2. Type `Something beautiful`  - it's fine to use predictive-text suggestions to help
3. Select the part of the second word: 'utiful'.
4. Press letter 'n'. 

TEST_SELECTION_BACKSPACE_MID:  Delete selected text that starts in the middle of a word.
1. Install the Keyman for iOS version associated with this PR for user testing.
2. Type `Something beautiful`  - it's fine to use predictive-text suggestions to help
3. Select the part of the second word: 'utiful'.
4. Press backspace.

TEST_GENERAL_USE:  Check that general usage of the iOS Keyman app acts as expected.

### CODE REVIEW

**Comment by bharanidharanj** [comment]
# Test Results

* **TEST_SELECTION_REPLACEMENT (PASSED):** Tested with the attached PR build (17.0.303-beta-test-11179) on an iPhone 13 Mobile device and here is my observation: 1. Typed 'Something beautiful'. 2. Selected the second word 'beautiful'. 3. Pressed letter 'B'. Verified that there is space between Something and B on the text input screen. Seems to be working fine.

**Comment by bharanidharanj** [comment]
# Test Results

* **TEST_SELECTION_BACKSPACE (FAILED):** Tested with the attached PR build (Keyman 17.0.303-beta-test-11179) on an iPhone 13 mobile device (iOS 17.4) and here is my observation: 1. Opened the Keyman In-app. 2. Typed 'Something beautiful' on the screen. 3. Selected the second word 'beautiful'. 4. Pressed the backspace key to delete and noticed that the selected word still appeared on the screen. It seems that the backspace key is not working on the selected text at all. Seems to be an issue.

**Comment by bharanidharanj** [comment]
# Test Results

* **TEST_GENERAL_USE (FAILED):** Tested with the attached PR build (Keyman 17.0.303-beta-test-11179) on an iPhone 13 mobile device and here is my observation: 1. Opened the Keyman In-app. 2. Noticed that after installing the Khmer Angkor keyboard, the corresponding dictionary was missing in the suggestion bar. 3. Installed four Keyman Keyboards using Settings menu. 5. Set Keyman as the system wide keyboard. Opened Safari browser. Switch to Keyman keyboard. Noticed that there is only one keyboard (EuroLatin (SIL)) is shown in the keyboard menu list. This appears to be an issue. 

..missing khmer_angkor dictionary in the suggestion bar

   


..missing other keyman keyboard in keyboard list

**Comment by jahorton** [comment]
> # Test Results
> 
>     * **TEST_GENERAL_USE (FAILED):** Tested with the attached PR build (Keyman 17.0.303-beta-test-11179) on an iPhone 13 mobile device and here is my observation: 1. Opened the Keyman In-app. 2. Noticed that after installing the Khmer Angkor keyboard, the corresponding dictionary was missing in the suggestion bar. 3. Installed four Keyman Keyboards using Settings menu. 5. Set Keyman as the system wide keyboard. Opened Safari browser. Switch to Keyman keyboard. Noticed that there is only one keyboard (EuroLatin (SIL)) is shown in the keyboard menu list. This appears to be an issue.
> 
> 
> ..missing khmer_angkor dictionary in the suggestion bar
> 
> 
> ..missing other keyman keyboard in keyboard list
> 

Huh, that's... useful test data to have, though the observed issues are unrelated to this PR.  I'll make a separate tracking issue.

**Comment by jahorton** [comment]
Whose bright idea was it at Apple to have `textDocumentProxy.insertText("")` - an empty string insert - do _nothing_ when there's selected text active?  😞 Any other actual string will nuke the selection... but not that.  So, yay, workaround time.

@keymanapp-test-bot retest all

I've fixed the error plaguing the second user-test now.

As for the third user test, please verify that "Allow Full Access" has been toggled - that would probably affect the system keyboard issue you saw.  As for the Khmer model not downloading... perhaps there was a connection hiccup?  Perhaps fully uninstall and reinstall the Keyman app, to be safe.

**Comment by bharanidharanj** [comment]
# Test Results

* **TEST_SELECTION_REPLACEMENT (PASSED):** Tested with the attached PR build (17.0.305-beta-test-11179) on an iPhone 13 Mobile (iOS 17.4) device and here is my observation: 1. Typed 'Something beautiful'. 2. Selected the second word 'beautiful'. 3. Pressed letter 'B'. Verified that there is space between Something and B on the text input screen. Seems to be working fine.

* **TEST_SELECTION_BACKSPACE (PASSED):** 1. Opened the Keyman In-app. 2. Typed 'Something beautiful' on the screen. 3. Selected the second word 'beautiful'. 4. Pressed the backspace key to delete the word 'beautiful' and verified that it was deleted and the preceding space remained unchanged. Seems to be working as expected.

**Comment by bharanidharanj** [comment]
# Test Results

* **TEST_GENERAL_USE (FAILED):** Retested with the attached PR build (Keyman 17.0.305-beta-test-11179) on an iPhone 13 Mobile device (iOS 17.4) and here is my observation: 1. Set Keyman as the system wide keyboard. 2. Opened Safari browser. 3. Switched to Keyman Keyboard. 4. Typed a text on the Search bar. 5. Locked the mobile device by pressing Lock key from the mobile. 6. Unlock the mobile. Here, I noticed that the Keyman Keyboard was no longer visible in the Keyboard panel. Seems to be an issue.

**Comment by jahorton** [comment]
> # Test Results
> 
> * **TEST_GENERAL_USE (FAILED):** Retested with the attached PR build (Keyman 17.0.305-beta-test-11179) on an iPhone 13 Mobile device (iOS 17.4) and here is my observation: 1. Set Keyman as the system wide keyboard. 2. Opened Safari browser. 3. Switched to Keyman Keyboard. 4. Typed a text on the Search bar. 5. Locked the mobile device by pressing Lock key from the mobile. 6. Unlock the mobile. Here, I noticed that the Keyman Keyboard was no longer visible in the Keyboard panel. Seems to be an issue.
>    

Interesting that we haven't seen this one until now.  There always have been odd interactions with that specific use case - see #2300 - and I can see the host-page getting reloaded / restored after locking and unlocking.  It's _temporarily_ blank for me, rather than permanently blank... but I'm also not sitting and waiting between locking & unlocking.  If something were to prevent the host-page reload, that could certainly cause the issue.  Of course, _getting evidence_ of that is an entirely different matter.

**Comment by jahorton** [comment]
I've created a separate issue corresponding to the blank-keyboard test result; this PR's changes affect nothing near the reported behavior.

**Comment by sgschantz** [comment]
@jahorton The behavior described in TEST_SELECTION_BACKSPACE appears to differ from the behavior for the Apple keyboard. Should a backspace be treated differently than typing a letter key if text is selected?

File: ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift [hunk 1]
```diff
@@ -347,18 +347,40 @@ open class InputViewController: UIInputViewController, KeymanWebDelegate {
...
-    if numCharsToDelete <= 0 || hasDeletedSelection {
+    if deleteSelection && newText == "" {
+      /*
+        if deleteSelection && newText == "", we have a backspace on
+        selected text.  Sadly, .insertText("")... does nothing.  Why, Apple!?
+       
+        The one silver lining:  Apple makes it impossible for users to select text
+        in a way that splits character clusters.
+       
+        So, we can just insert something that won't combine, like a ZWNJ, and then delete it.
```
Comment (sgschantz) [inline ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift:366]: Are you saying that you are working around the fact that insertText of the empty string does not cause the selection to be replaced by doing an insert of something invisible that we can then delete? Does this only happen for backspace?

It sounds like you are detecting this backspace case (based on line 360) by saying we have an empty string to insert and the text selection is greater than zero. Is that only true for backspace? 

This comment is hard to follow because the problem is identified, then the silver lining is mentioned and then the workaround. Also, it would be nice to keep the emotion or sarcasm out of the code as it can distract from the technical content -- even if the technical content is correct. The same is true for comments in issues or PRs. This is not productive for our relationships with our vendors.

File: ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift [hunk 2]
```diff
@@ -347,18 +347,40 @@ open class InputViewController: UIInputViewController, KeymanWebDelegate {
...
+      let afterManipulation = textDocumentProxy.documentContextBeforeInput ?? ""
+      
+      // For good measure, a canary to signal if our selected-text backspace handling
+      // goes awry.
+      if beforeManipulation != afterManipulation {
+        os_log(.error, log: KeymanEngineLogger.engine, "Could not cleanly execute backspace for selected text")
+      }
+      sendContextUpdate()
+      return
+    } else if numCharsToDelete <= 0 || deleteSelection {
```
Comment (sgschantz) [inline ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift:382]: This method was already long, and the new additions would be easier to follow and maintain if they were encapsulated in their own methods. If someone is reading this code and not concerned with the edge case, then it would be helpful for the edge case to be contained separately. Some of the comments could also be reduced if the new methods were named in a way that indicated what their purpose was.

**Comment by jahorton** [comment]
> @jahorton The behavior described in TEST_SELECTION_BACKSPACE appears to differ from the behavior for the Apple keyboard. Should a backspace be treated differently than typing a letter key if text is selected?

That's because Apple does fancy handling with spaces.  Type `temptext` and select just the `text` part, then hit backspace.  The `p` of `temp` will remain.  Interestingly... wait, the `p` remains with our current implementation's use of `.deleteBackward()`, too.

The original issue is about how we _don't_ want that space deleted if we're replacing text - use of `.deleteBackward()` does nuke that space, after all.  So, preserve the space if _replacing_ text, not if _removing_ text - that seems to be what you're saying there?

**Comment by jahorton** [comment]
Before I forget, that makes things pretty darn tricky:  the internal Web engine doesn't know or have Apple's default handling for that scenario.  We're only safe to delete that space if we can ensure the edit is properly synchronized with the internal Web engine.  At present... I'm pretty sure that would trigger a context reset, and we don't want that.

Due to limitations, and how late we are in the release cycle, I think we need to break from how Apple does it - otherwise, we risk triggering context-desyncs, which are quite possibly the mobile app's nastiest class of errors.

File: ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift (hunk 1 above)
Comment (jahorton) [reply ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift:366]: > Are you saying that you are working around the fact that insertText of the empty string does not cause the selection to be replaced by doing an insert of something invisible that we can then delete?

Yes.  That's exactly what it's saying.

> Does this only happen for backspace?
>
> It sounds like you are detecting this backspace case (based on line 360) by saying we have an empty string to insert and the text selection is greater than zero. Is that only true for backspace?

It's theoretically possible for other backspace-like keys to be defined, I suppose.  Changes to keyboard options don't trigger `insertText` unless they also _actually_ manipulate text.

Suppose the following:

```swift
self.insertText(keymanWeb , /*numCharsToDelete:*/ 2, /*newText*/ "a")
```

If text is selected that immediately follows a space, we need to _preserve_ that space in order to properly match how many characters the internal Web engine has said to delete.  (In fact... we need to preserve it regardless in order to stay synch'd with the internal engine.)  We have to do the deletions _before_ our text-insertions, too.  

So... I think this needs to happen this way _regardless_ of whether or not we're inserting text down the line - otherwise, we run the risk of desynchronizing.  If `temptext` is the context and `text` is selected, as this code currently is, at the time of this comment, the first char to delete will be silenced due to erasing the selected text while _not_ deleting the preceding `p`.

So... thanks for raising the point that helped me catch this.

**Comment by sgschantz** [comment]
> Before I forget, that makes things pretty darn tricky: the internal Web engine doesn't know or have Apple's default handling for that scenario. We're only safe to delete that space if we can ensure the edit is properly synchronized with the internal Web engine. At present... I'm pretty sure that would trigger a context reset, and we don't want that.
> 
> Due to limitations, and how late we are in the release cycle, I think we need to break from how Apple does it - otherwise, we risk triggering context-desyncs, which are quite possibly the mobile app's nastiest class of errors.

I checked for consistency with Apple's keyboard because I didn't have a strong intuitive feel for whether the space should be deleted when backspacing with a word selected. Just for comparison, I looked at Pages and found the same behavior. Then I looked at TextEdit and saw that the backspace does not remove the space preceeding the selected word. Should this be the same for desktop and mobile? I don't know.

It seems more consistent to say, in both cases, that the key typed should be applied after deleting the selected text, meaning the backspace will remove the preceeding space. However, when typing in a text editor on a desktop platform, if I select a word and hit backspace, I don't think I want that. In that moment, I expect the backspace key to act like a delete key (which is what the key is actually called on the Mac, anyway).

So I agree -- we should go with whatever is safest for now. We can map out all the scenarios in the future when we deal in depth with selected text behavior.

**Comment by jahorton** [comment]
@keymanapp-test-bot retest TEST_SELECTION_REPLACEMENT TEST_SELECTION_BACKSPACE

The selection-text removal code has been changed, so it's best to double-check that the new pattern holds for all cases.  A couple of new unit tests have been added, now that new potential "wrinkles" have been identified.

Ideally, I should also add an extra test case that includes use of a key that deletes more than one character yet isn't a backspace - that would be a _real_ stress-test.

**Comment by bharanidharanj** [comment]
# Test Results

* **TEST_SELECTION_REPLACEMENT (PASSED):** Tested with the attached PR build (Keyman 17.0.307-beta-test) on an iPhone mobile device and here is my observation: 1. Typed 'Something beautiful'. 2. Selected the second word 'beautiful'. Presse the letter B. 3. Verified that there is a space appears between Something and B. 
* **TEST_SELECTION_BACKSPACE (PASSED):** 1. Verified that a selected word can be deleted by pressing the backspace key, with no further edits. 
* **TEST_SELECTION_REPLACEMENT_MID (PASSED):** 1. Typed 'Something beautiful'. 2. Selected a part of the second word 'utiful'. 3. Pressed letter 'n. Verified that it showed 'Something bean' on the text input screen. 
* **TEST_SELECTION_BACKSPACE_MID (PASSED):** 1. Typed 'Something beautiful'. 2. Selected a part of the second word 'utiful'. 3. Pressed the backspace key. Verified that it deleted the selected word. It showed 'Something bea'.

**Comment by bharanidharanj** [comment]
@jahorton It appears that the last test, TEST_GENERAL_USE, seems to be resolved in Keyman 17.0.307-beta-test-11179 and is functioning correctly. Could you please reassign this test? Thanks.

**Comment by mcdurdin** [comment]
> Whose bright idea was it at Apple to have `textDocumentProxy.insertText("")` - an empty string insert - do _nothing_ when there's selected text active? 😞 Any other actual string will nuke the selection... but not that. So, yay, workaround time.

Have we reported this to Apple as a bug? May be worth doing so.

File: ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift [hunk 3]
```diff
@@ -333,6 +333,60 @@ open class InputViewController: UIInputViewController, KeymanWebDelegate {
...
+
+      /*
+        Apple has some special nuances to its text deletion that our internal
+        Web engine does not emulate.
+       
+        .deleteBackward() behaviors:
+        - If there is selected text immediately following a space (U+0020),
+          it will delete that space IN ADDITION to the selected text.
+        - If there is selected text that starts mid-word, it will NOT delete
+          the preceding character.
```
Comment (mcdurdin) [inline ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift:354]: This seems to be just one special case for space preceding a selection, not two special cases. The way this is written it looks like there are two special cases.

Remember that our policy on backspace with selected text is to delete the selection, but not the preceding character (#9073).

File: ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift [hunk 4]
```diff
@@ -333,6 +333,60 @@ open class InputViewController: UIInputViewController, KeymanWebDelegate {
...
+       
+        Compare to Web:  Web states an exact number of characters to delete
+        before the start of the currently-selected range... and it does this
+        completely unaware of the nuances listed above for .deleteBackward().
+        Keyman keyboard rules are likewise unaware of Apple's nuances.
```
Comment (mcdurdin) [inline ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift:359]: This is a bit of a side-track. 

```suggestion
        Our policy (#9073) on handling the backspace key when there is a 
        text selection is to just delete the selection. We have to override
        the special case of space being deleted by .deleteBackward() ourselves.
```

File: ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift [hunk 5]
```diff
@@ -333,6 +333,60 @@ open class InputViewController: UIInputViewController, KeymanWebDelegate {
...
+       
+        In order to maintain proper synchronization between app context and
+        internal Web-engine context, we need to force selected-text deletion
+        to NEVER delete preceding spaces.  Any attempts to adjust and include
+        the aforementioned nuance will need considerable design work to "get
+        right" due to the risk for adverse affects with Keyman keyboard rules.
```
Comment (mcdurdin) [inline ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift:365]:
```suggestion
```

I don't think this comment is needed. Just the technical background.

File: ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift [hunk 6]
```diff
@@ -333,6 +333,60 @@ open class InputViewController: UIInputViewController, KeymanWebDelegate {
...
+       
+        .insertText() is great for this... when the string isn't empty.  If
+        it is, well, "sorry, out of luck."  That said, we can just insert
+        something that won't combine, like a ZWNJ, and then delete it.
+       
+        The silver lining:  Apple makes it impossible for users to select text
+        in a way that splits character clusters.  This implies that it's
+        impossible for an inserted ZWNJ to combine with existing context, 
+        making this operation safe.
```
Comment (mcdurdin) [inline ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift:374]:
```suggestion
        As .insertText() does not delete the selection if the string to
        be inserted is empty, we insert something that won't combine, 
        like a ZWNJ, and then delete it.
       
        iOS does not allow users to select text in a way that splits 
        character clusters.  This implies that it's impossible for an
        inserted ZWNJ to combine with existing context, making this 
        operation safe.
```

File: ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift (hunk 3 above)
Comment (mcdurdin) [reply ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift:354]:
```suggestion
        We have a problem to resolve here: we cannot simply delete the selection 
        with either .deleteBackward() or .insertText(""):
        
        - If there is selected text immediately following a space (U+0020),
          .deleteBackward() will delete that space IN ADDITION to the selected text.
        - Unlike .insertText("-any-string-here"), .insertText("") does nothing -- 
          it does not replace the selection with the new string.
```

File: ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift [hunk 7]
```diff
@@ -347,18 +401,11 @@ open class InputViewController: UIInputViewController, KeymanWebDelegate {
       perform(#selector(self.enableInputClickSound), with: nil, afterDelay: 0.1)
     }
 
-    var hasDeletedSelection = false
+    // `true` if there was selected text to be deleted
+    let deletedSelection = self.deleteSelection()
```
Comment (mcdurdin) [inline ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift:405]: We only need to do this call to `deleteSelection()` if we (a) there is a selection, and (b) we have no new text to insert, right? We really don't want to jump through this hoop if we don't have to.

File: ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift [hunk 8]
```diff
@@ -347,18 +401,11 @@ open class InputViewController: UIInputViewController, KeymanWebDelegate {
...
-    if let selected = textDocumentProxy.selectedText {
-      if selected.count > 0 {
-        textDocumentProxy.deleteBackward()
-        hasDeletedSelection = true
-      }
-    }
-
-    if numCharsToDelete <= 0 || hasDeletedSelection {
+    if numCharsToDelete <= 0 || deletedSelection {
       textDocumentProxy.insertText(newText)
```
Comment (mcdurdin) [inline ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift:408]: This seems that if (a) we have no selection, and (b) no deletion, and (c) text to insert, then it won't insert the new text?

**Review by mcdurdin — CHANGES_REQUESTED** [review]

File: ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift (hunk 1 above)
Comment (mcdurdin) [reply ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift]: > Also, it would be nice to keep the emotion or sarcasm out of the code as it can distract from the technical content -- even if the technical content is correct.

I have at times been guilty of this. It's not a good practice, I agree. Frustration can be expressed in Slack ;-) and we'll keep the code comments, issues, and PRs focused on technical.

> So... I think this needs to happen this way _regardless_ of whether or not we're inserting text down the line - otherwise, we run the risk of desynchronizing. If `temptext` is the context and `text` is selected, as this code currently is, at the time of this comment, the first char to delete will be silenced due to erasing the selected text while _not_ deleting the preceding `p`.

.insertText() will delete selection if the string is not empty, right? And in that scenario I don't think it deletes the space prior to the selection. AFAICT, we only need to jump through the hoop when we are deleting selection with no insertion (generally, backspace)

File: ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift (hunk 5 above)
Comment (jahorton) [reply ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift:365]: ... this _is_ technical background, just from a different starting point.  As the primary maintainer of the Web/mobile-app interface, _this_ was the reason I made the change, after all - I wasn't even aware of the decision from #9073 when I made this.  It just so happens that both reasons align.

**Comment by jahorton** [comment]
> > Whose bright idea was it at Apple to have `textDocumentProxy.insertText("")` - an empty string insert - do _nothing_ when there's selected text active? 😞 Any other actual string will nuke the selection... but not that. So, yay, workaround time.
> 
> Have we reported this to Apple as a bug? May be worth doing so.

I have now submitted FB13734388 in regard to this.

File: ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift (hunk 8 above)
Comment (jahorton) [reply ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift:408]: No, it just comes later, after deleting preceding text.  In the case that no such deletions are needed, this section allows us to return early before all of the delete-left handling kicks in.  No edits were made near the later code block, so it doesn't show up in the default code-review code-windows.

Comment (jahorton) [reply ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift:408]: But... I think there is something off here; we probably shouldn't condition on the `deletedSelection` bit.

Comment (jahorton) [reply ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift:408]: Oh right.  If we do have a selection, we currently act as if the context is nil and thus can't back delete, so it's probably a moot point.

File: ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift (hunk 7 above)
Comment (jahorton) [reply ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift:405]: That's why the guard exists at the start of `deleteSelection`.  I placed it within `deleteSelection` rather than outside it.

File: ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift [hunk 9]
```diff
@@ -333,6 +333,60 @@ open class InputViewController: UIInputViewController, KeymanWebDelegate {
       }
     }
   }
+  
+  func deleteSelection() -> Bool {
+    if let selected = textDocumentProxy.selectedText, selected.count > 0 {
```
Comment (jahorton) [inline ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift:338]: This guard ensures we only try to delete selected text only if there actually _is_ selected text - a selection exists with length > 0.  If it doesn't exist, we exit and indicate that no deletion happened.

File: ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift (hunk 8 above)
Comment (jahorton) [reply ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift:405]: The "enforcer" for "selected text" = "empty context", at least within the internal Web engine:

https://github.com/keymanapp/keyman/blob/cffabd719edf6c32297120e2a1ab5639ddc50faa/common/web/keyboard-processor/src/text/kbdInterface.ts#L302-L306

**Review by mcdurdin — APPROVED** [review]
LGTM
